src/libraries/System.IO.FileSystem.AccessControl/src/System/IO/MemoryMappedFiles/MemoryMappedFileAcl.cs

```diff
@@ -0,0 +1,101 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Reflection;
+using System.Security.AccessControl;
+
+namespace System.IO.MemoryMappedFiles
+{
+    public static class MemoryMappedFileAcl
+    {
+        private static Type MemoryMappedFileType { get; } = typeof(MemoryMappedFile);
+
+        public static unsafe MemoryMappedFile CreateFromFile(
+            FileStream fileStream,
+            string mapName,
+            long capacity,
+            MemoryMappedFileAccess access,
+            MemoryMappedFileSecurity memoryMappedFileSecurity,
+            HandleInheritability inheritability,
+            bool leaveOpen)
+        {
+            fixed (byte* pSecurityDescriptor = memoryMappedFileSecurity.GetSecurityDescriptorBinaryForm())
+            {
+                return GetMemoryMappedFileWithTypeMemberInvocation("CreateFromFile",
```

Comment:
Can we add public APIs that take a handle instead of using this reflection hack?

Reply:
We can look into making the [private constructors](https://source.dot.net/#System.IO.MemoryMappedFiles/System/IO/MemoryMappedFiles/MemoryMappedFile.cs,28) public. If we do, we would want to include the core logic in both assemblies to avoid code duplication (to reduce the risk of mismatching with the other APIs).

Reply:
What about adding a setter to the [public SafeMemoryMappedFileHandle property](https://source.dot.net/#System.IO.MemoryMappedFiles/System/IO/MemoryMappedFiles/MemoryMappedFile.cs,482)?

The reason why I suggest this alternative is because the ACL APIs we added for Mutex, Semaphore and EventWaitHandle all have that setter, and none have a public constructor that receives a handle.

Since we had the public setter, we were able to:
1. Create the handle with the appropriate security.
2. Create the object.
3. Switch its inner handle with the one we created.

[Example for EventWaitHandle](https://source.dot.net/#System.Threading.AccessControl/System/Threading/EventWaitHandleAcl.cs,64).

Reply:
> Mutex, Semaphore and EventWaitHandle

WaitHandle that all these inherit from is a thin wrapper around the underlying OS handle. WaitHandle has just the safe handle field and nothing else. It makes it reasonable to have a setter for the safe handle.

`MemoryMappedFile` is not a thin wrapper like that. It has multiple fields that better stay in sync. It would be hard to reason about the behavior of the public setter given this.